util/merkle/src/nibbleslice.rs

```diff
@@ -0,0 +1,300 @@
+    /// noting whether it `is_leaf`.
+    pub fn encoded_leftmost(&self, n: usize) -> ElasticArray36<u8> {
+        let l = min(self.len(), n);
+        let mut r = ElasticArray36::new();
+        let mut i = l % 2;
+        r.push(if i == 1 {
+            0x10 + self.at(0)
+        } else {
+            0
+        });
+        while i < l {
+            r.push(self.at(i) * 16 + self.at(i + 1));
+            i += 2;
+        }
+        r
+    }
+}
+
+impl<'a> PartialEq for NibbleSlice<'a> {
+    fn eq(&self, them: &Self) -> bool {
+        self.len() == them.len() && self.starts_with(them)
+    }
+}
+
+impl<'a> PartialOrd for NibbleSlice<'a> {
+    fn partial_cmp(&self, them: &Self) -> Option<Ordering> {
+        let s = min(self.len(), them.len());
+        let mut i = 0usize;
+        while i < s {
+            match self.at(i).partial_cmp(&them.at(i)).unwrap() {
+                Ordering::Less => return Some(Ordering::Less),
+                Ordering::Greater => return Some(Ordering::Greater),
+                _ => i += 1,
+            }
+        }
+        self.len().partial_cmp(&them.len())
+    }
+}
+
+impl<'a> fmt::Debug for NibbleSlice<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for i in 0..self.len() {
+            match i {
+                0 => write!(f, "{:01x}", self.at(i))?,
+                _ => write!(f, "'{:01x}", self.at(i))?,
+            }
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::NibbleSlice;
+    use elastic_array::ElasticArray36;
+    static D: &'static [u8; 3] = &[0x01u8, 0x23, 0x45];
+
+    #[test]
+    fn basics() {
+        let n = NibbleSlice::new(D);
+        assert_eq!(n.len(), 6);
+        assert!(!n.is_empty());
+
+        let n = NibbleSlice::new_offset(D, 6);
+        assert!(n.is_empty());
+
+        let n = NibbleSlice::new_offset(D, 3);
+        assert_eq!(n.len(), 3);
+        for i in 0..3 {
+            assert_eq!(n.at(i), i as u8 + 3);
+        }
+    }
+
+    #[test]
+    fn mid() {
+        let n = NibbleSlice::new(D);
+        let m = n.mid(2);
+        for i in 0..4 {
+            assert_eq!(m.at(i), i as u8 + 2);
+        }
+        let m = n.mid(3);
+        for i in 0..3 {
+            assert_eq!(m.at(i), i as u8 + 3);
+        }
+    }
+
+    #[test]
+    fn encoded() {
+        let n = NibbleSlice::new(D);
+        assert_eq!(n.encoded(), ElasticArray36::from_slice(&[0x00, 0x01, 0x23, 0x45]));
+        assert_eq!(n.mid(1).encoded(), ElasticArray36::from_slice(&[0x11, 0x23, 0x45]));
+    }
+
+    #[test]
+    fn from_encoded() {
+        let n = NibbleSlice::new(D);
+        assert_eq!(n, NibbleSlice::from_encoded(&[0x00, 0x01, 0x23, 0x45]));
+        assert_eq!(n, NibbleSlice::from_encoded(&[0x20, 0x01, 0x23, 0x45]));
+        assert_eq!(n.mid(1), NibbleSlice::from_encoded(&[0x11, 0x23, 0x45]));
+        assert_eq!(n.mid(1), NibbleSlice::from_encoded(&[0x31, 0x23, 0x45]));
```

Comment:
Are both `31 23 45` and `11 23 45` valid encoding of same value?
I think there should be only one valid encoding for consistent hash value.

Reply:
That was wrong test case. I change it correctly